tests/api/test_ratelimiting.py

```diff
@@ -1,39 +1,111 @@
-from synapse.api.ratelimiting import Ratelimiter
+from synapse.api.ratelimiting import LimitExceededError, Ratelimiter
 
 from tests import unittest
 
 
 class TestRatelimiter(unittest.TestCase):
     def test_allowed(self):
-        limiter = Ratelimiter()
-        allowed, time_allowed = limiter.can_do_action(
-            key="test_id", time_now_s=0, rate_hz=0.1, burst_count=1
-        )
+        limiter = Ratelimiter(clock=None, rate_hz=0.1, burst_count=1)
+        allowed, time_allowed = limiter.can_do_action(key="test_id", _time_now_s=0)
         self.assertTrue(allowed)
         self.assertEquals(10.0, time_allowed)
 
-        allowed, time_allowed = limiter.can_do_action(
-            key="test_id", time_now_s=5, rate_hz=0.1, burst_count=1
-        )
+        allowed, time_allowed = limiter.can_do_action(key="test_id", _time_now_s=5)
         self.assertFalse(allowed)
         self.assertEquals(10.0, time_allowed)
 
+        allowed, time_allowed = limiter.can_do_action(key="test_id", _time_now_s=10)
+        self.assertTrue(allowed)
+        self.assertEquals(20.0, time_allowed)
+
+    def test_allowed_via_ratelimit(self):
+        limiter = Ratelimiter(clock=None, rate_hz=0.1, burst_count=1)
+
+        # Shouldn't raise
+        limiter.ratelimit(key="test_id", _time_now_s=0)
+
+        # Should raise
+        self.assertRaises(
+            LimitExceededError, limiter.ratelimit, key="test_id", _time_now_s=5,
+        )
+
+        # Shouldn't raise
+        limiter.ratelimit(key="test_id", _time_now_s=10)
+
+    def test_allowed_by_overriding_parameters(self):
+        """Test that we can override options of a Ratelimiter that would otherwise fail
+        an action
+        """
+        # Create a Ratelimiter with a very low allowed rate_hz and burst_count
+        limiter = Ratelimiter(clock=None, rate_hz=0.1, burst_count=1)
+
+        # First attempt should be allowed
+        time_now = 0
+        expected_allowed = 10.0
+
+        # Shouldn't raise
+        limiter.ratelimit(("test_id",), _time_now_s=time_now, update=False)
+
         allowed, time_allowed = limiter.can_do_action(
-            key="test_id", time_now_s=10, rate_hz=0.1, burst_count=1
+            key=("test_id",), _time_now_s=time_now,
         )
         self.assertTrue(allowed)
-        self.assertEquals(20.0, time_allowed)
+        self.assertEquals(expected_allowed, time_allowed)
+
+        # Second attempt, 1s later, will fail
+        time_now = 1
+        expected_allowed = 10.0
+
+        # We expect a LimitExceededError to be raised
+        try:
+            limiter.ratelimit(("test_id",), _time_now_s=time_now, update=False)
+
+            # We shouldn't reach here
+            self.assertTrue(False, "LimitExceededError was not raised")
+        except LimitExceededError as e:
+            self.assertEquals(e.retry_after_ms / 1000, expected_allowed - time_now)
```

Comment:
This would be nicer using the context manager style:

```python
with self.assertRaises(LimiteExceedError) as e:
    limiter.rateLimit(...)

# Use e below
self.assertEquals(...)
```

Reply:
Was not aware this was supported by `self.assertRaises`. Thanks for the tip!